Support ClickHouse DateTime(<timezone>) data type

Hello.

Trino 389.

Currently Trino doesn't support `DateTime(<timezone>)` type in Clickhouse, and, even worse, `unsupported-type-handling=CONVERT_TO_VARCHAR` also fails to handle it. When I turn it on, I get the following error: 

```Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}```

<details>
  <summary>Full stacktrace</summary>

```
2022-08-08T10:04:37.180Z       ERROR   SplitRunner-42-139      io.trino.execution.executor.TaskExecutor        Error processing Split 20220808_100436_00001_23krj.2.0.0-0 SystemSplit{addresses=[172.25.0.12:8080]} (start = 3.6987939359458916E10, wall = 99 ms, cpu = 0 ms, wait = 0 ms, calls = 1)
 com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2055)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4863)
        at io.trino.collect.cache.EvictableCache.get(EvictableCache.java:108)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.getTableMetadata(DefaultJdbcMetadata.java:629)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.lambda$listTableColumns$10(DefaultJdbcMetadata.java:672)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.listTableColumns(DefaultJdbcMetadata.java:672)
        at io.trino.spi.connector.ConnectorMetadata.streamTableColumns(ConnectorMetadata.java:292)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:318)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:85)
        at io.trino.metadata.MetadataManager.listTableColumns(MetadataManager.java:547)
        at io.trino.metadata.MetadataListing.listTableColumns(MetadataListing.java:166)
        at io.trino.connector.system.jdbc.ColumnJdbcTable.cursor(ColumnJdbcTable.java:263)
        at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
        at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
        at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
        at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
        at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
        at io.trino.operator.Driver.processInternal(Driver.java:410)
        at io.trino.operator.Driver.lambda$process$10(Driver.java:313)
        at io.trino.operator.Driver.tryWithLock(Driver.java:698)
        at io.trino.operator.Driver.process(Driver.java:305)
        at io.trino.operator.Driver.processForDuration(Driver.java:276)
        at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
        at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
        at io.trino.$gen.Trino_389____20220808_100410_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
 Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:72)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4868)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
        ... 47 more
 Caused by: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.base.Verify.verify(Verify.java:417)
        at io.trino.plugin.jdbc.BaseJdbcClient.getColumns(BaseJdbcClient.java:311)
        at io.trino.plugin.jdbc.ForwardingJdbcClient.getColumns(ForwardingJdbcClient.java:96)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$getColumns$5(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:35)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.getColumns(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:66)
        ... 55 more


 2022-08-08T10:04:37.197Z       ERROR   remote-task-callback-3  io.trino.execution.scheduler.PipelinedStageExecution    Pipelined stage execution for stage 20220808_100436_00001_23krj.2 failed
 com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2055)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4863)
        at io.trino.collect.cache.EvictableCache.get(EvictableCache.java:108)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.getTableMetadata(DefaultJdbcMetadata.java:629)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.lambda$listTableColumns$10(DefaultJdbcMetadata.java:672)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.listTableColumns(DefaultJdbcMetadata.java:672)
        at io.trino.spi.connector.ConnectorMetadata.streamTableColumns(ConnectorMetadata.java:292)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:318)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:85)
        at io.trino.metadata.MetadataManager.listTableColumns(MetadataManager.java:547)
        at io.trino.metadata.MetadataListing.listTableColumns(MetadataListing.java:166)
        at io.trino.connector.system.jdbc.ColumnJdbcTable.cursor(ColumnJdbcTable.java:263)
        at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
        at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
        at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
        at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
        at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
        at io.trino.operator.Driver.processInternal(Driver.java:410)
        at io.trino.operator.Driver.lambda$process$10(Driver.java:313)
        at io.trino.operator.Driver.tryWithLock(Driver.java:698)
        at io.trino.operator.Driver.process(Driver.java:305)
        at io.trino.operator.Driver.processForDuration(Driver.java:276)
        at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
        at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
        at io.trino.$gen.Trino_389____20220808_100410_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
 Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:72)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4868)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
        ... 47 more
 Caused by: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.base.Verify.verify(Verify.java:417)
        at io.trino.plugin.jdbc.BaseJdbcClient.getColumns(BaseJdbcClient.java:311)
        at io.trino.plugin.jdbc.ForwardingJdbcClient.getColumns(ForwardingJdbcClient.java:96)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$getColumns$5(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:35)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.getColumns(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:66)
        ... 55 more


 2022-08-08T10:04:37.201Z       ERROR   stage-scheduler io.trino.execution.scheduler.SqlQueryScheduler  Failure in distributed stage for query 20220808_100436_00001_23krj
 com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2055)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4863)
        at io.trino.collect.cache.EvictableCache.get(EvictableCache.java:108)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.getTableMetadata(DefaultJdbcMetadata.java:629)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.lambda$listTableColumns$10(DefaultJdbcMetadata.java:672)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.listTableColumns(DefaultJdbcMetadata.java:672)
        at io.trino.spi.connector.ConnectorMetadata.streamTableColumns(ConnectorMetadata.java:292)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:318)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:85)
        at io.trino.metadata.MetadataManager.listTableColumns(MetadataManager.java:547)
        at io.trino.metadata.MetadataListing.listTableColumns(MetadataListing.java:166)
        at io.trino.connector.system.jdbc.ColumnJdbcTable.cursor(ColumnJdbcTable.java:263)
        at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
        at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
        at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
        at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
        at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
        at io.trino.operator.Driver.processInternal(Driver.java:410)
        at io.trino.operator.Driver.lambda$process$10(Driver.java:313)
        at io.trino.operator.Driver.tryWithLock(Driver.java:698)
        at io.trino.operator.Driver.process(Driver.java:305)
        at io.trino.operator.Driver.processForDuration(Driver.java:276)
        at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
        at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
        at io.trino.$gen.Trino_389____20220808_100410_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
 Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:72)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4868)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
        ... 47 more
 Caused by: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.base.Verify.verify(Verify.java:417)
        at io.trino.plugin.jdbc.BaseJdbcClient.getColumns(BaseJdbcClient.java:311)
        at io.trino.plugin.jdbc.ForwardingJdbcClient.getColumns(ForwardingJdbcClient.java:96)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$getColumns$5(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:35)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.getColumns(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:66)
        ... 55 more


 2022-08-08T10:04:37.211Z       ERROR   stage-scheduler io.trino.execution.StageStateMachine    Stage 20220808_100436_00001_23krj.2 failed
 com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2055)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4863)
        at io.trino.collect.cache.EvictableCache.get(EvictableCache.java:108)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.getTableMetadata(DefaultJdbcMetadata.java:629)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.lambda$listTableColumns$10(DefaultJdbcMetadata.java:672)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.trino.plugin.jdbc.DefaultJdbcMetadata.listTableColumns(DefaultJdbcMetadata.java:672)
        at io.trino.spi.connector.ConnectorMetadata.streamTableColumns(ConnectorMetadata.java:292)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:318)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.streamTableColumns(ClassLoaderSafeConnectorMetadata.java:85)
        at io.trino.metadata.MetadataManager.listTableColumns(MetadataManager.java:547)
        at io.trino.metadata.MetadataListing.listTableColumns(MetadataListing.java:166)
        at io.trino.connector.system.jdbc.ColumnJdbcTable.cursor(ColumnJdbcTable.java:263)
        at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:129)
        at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
        at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
        at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:108)
        at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:268)
        at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:196)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
        at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
        at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
        at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
        at io.trino.operator.Driver.processInternal(Driver.java:410)
        at io.trino.operator.Driver.lambda$process$10(Driver.java:313)
        at io.trino.operator.Driver.tryWithLock(Driver.java:698)
        at io.trino.operator.Driver.process(Driver.java:305)
        at io.trino.operator.Driver.processForDuration(Driver.java:276)
        at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
        at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:488)
        at io.trino.$gen.Trino_389____20220808_100410_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
 Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:72)
        at io.trino.plugin.jdbc.CachingJdbcClient.get(CachingJdbcClient.java:752)
        at io.trino.plugin.jdbc.CachingJdbcClient.getColumns(CachingJdbcClient.java:165)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4868)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
        ... 47 more
 Caused by: com.google.common.base.VerifyException: Unsupported type handling is set to CONVERT_TO_VARCHAR, but toColumnMapping() returned empty for JdbcTypeHandle{jdbcType=2014, jdbcTypeName=DateTime('Europe/Moscow'), columnSize=Optional[29], decimalDigits=Optional[0]}
        at com.google.common.base.Verify.verify(Verify.java:417)
        at io.trino.plugin.jdbc.BaseJdbcClient.getColumns(BaseJdbcClient.java:311)
        at io.trino.plugin.jdbc.ForwardingJdbcClient.getColumns(ForwardingJdbcClient.java:96)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$getColumns$5(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:35)
        at io.trino.plugin.jdbc.jmx.StatisticsAwareJdbcClient.getColumns(StatisticsAwareJdbcClient.java:116)
        at io.trino.plugin.jdbc.CachingJdbcClient.lambda$getColumns$3(CachingJdbcClient.java:165)
        at io.trino.collect.cache.EmptyCache.get(EmptyCache.java:66)
        ... 55 more

```
  
</details>


Actually it would be great to add support of the `DateTime(<timezone>)` type, but currently it would be nice to fix converting to varchar.

Thank you.

--- a/docs/src/main/sphinx/connector/clickhouse.rst
+++ b/docs/src/main/sphinx/connector/clickhouse.rst
@@ -238,8 +238,8 @@ to the following table:
   * - ``Date``
     - ``DATE``
     -
-  * - ``DateTime``
-    - ``TIMESTAMP(0)``
+  * - ``DateTime[(timezone)]``
+    - ``TIMESTAMP(0) [WITH TIME ZONE]``
     -
   * - ``IPv4``
     - ``IPADDRESS``

--- a/docs/src/main/sphinx/connector/clickhouse.rst
+++ b/docs/src/main/sphinx/connector/clickhouse.rst
@@ -239,7 +239,7 @@ to the following table:
     - ``DATE``
     -
   * - ``DateTime``
-    - ``TIMESTAMP``
+    - ``TIMESTAMP(0)``
     -
   * - ``IPv4``
     - ``IPADDRESS``

--- a/plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java
+++ b/plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java
@@ -35,6 +35,7 @@ import io.trino.plugin.jdbc.JdbcColumnHandle;
 import io.trino.plugin.jdbc.JdbcExpression;
 import io.trino.plugin.jdbc.JdbcTableHandle;
 import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.plugin.jdbc.LongReadFunction;
 import io.trino.plugin.jdbc.LongWriteFunction;
 import io.trino.plugin.jdbc.ObjectWriteFunction;
 import io.trino.plugin.jdbc.QueryBuilder;
@@ -59,6 +60,7 @@ import io.trino.spi.type.DecimalType;
 import io.trino.spi.type.Decimals;
 import io.trino.spi.type.Int128;
 import io.trino.spi.type.StandardTypes;
+import io.trino.spi.type.TimeZoneKey;
 import io.trino.spi.type.Type;
 import io.trino.spi.type.TypeManager;
 import io.trino.spi.type.TypeSignature;
@@ -79,8 +81,10 @@ import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Types;
+import java.time.Instant;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -143,6 +147,9 @@ import static io.trino.spi.StandardErrorCode.INVALID_TABLE_PROPERTY;
 import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
 import static io.trino.spi.type.BigintType.BIGINT;
 import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.trino.spi.type.DateTimeEncoding.unpackMillisUtc;
+import static io.trino.spi.type.DateTimeEncoding.unpackZoneKey;
 import static io.trino.spi.type.DateType.DATE;
 import static io.trino.spi.type.DecimalType.createDecimalType;
 import static io.trino.spi.type.DoubleType.DOUBLE;
@@ -151,6 +158,7 @@ import static io.trino.spi.type.RealType.REAL;
 import static io.trino.spi.type.SmallintType.SMALLINT;
 import static io.trino.spi.type.TimestampType.TIMESTAMP_MILLIS;
 import static io.trino.spi.type.TimestampType.TIMESTAMP_SECONDS;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_SECONDS;
 import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
 import static io.trino.spi.type.Timestamps.NANOSECONDS_PER_MICROSECOND;
 import static io.trino.spi.type.TinyintType.TINYINT;
@@ -588,6 +596,7 @@ public class ClickHouseClient
 
             case Types.TIMESTAMP:
                 if (columnDataType == ClickHouseDataType.DateTime) {
+                    // ClickHouse DateTime does not have sub-second precision
                     verify(typeHandle.getRequiredDecimalDigits() == 0, "Expected 0 as timestamp precision, but got %s", typeHandle.getRequiredDecimalDigits());
                     return Optional.of(ColumnMapping.longMapping(
                             TIMESTAMP_SECONDS,
@@ -596,6 +605,16 @@ public class ClickHouseClient
                 }
                 // TODO (https://github.com/trinodb/trino/issues/10537) Add support for Datetime64 type
                 return Optional.of(timestampColumnMappingUsingSqlTimestampWithRounding(TIMESTAMP_MILLIS));
+
+            case Types.TIMESTAMP_WITH_TIMEZONE:
+                if (columnDataType == ClickHouseDataType.DateTime) {
+                    // ClickHouse DateTime does not have sub-second precision
+                    verify(typeHandle.getRequiredDecimalDigits() == 0, "Expected 0 as timestamp with time zone precision, but got %s", typeHandle.getRequiredDecimalDigits());
+                    return Optional.of(ColumnMapping.longMapping(
+                            TIMESTAMP_TZ_SECONDS,
+                            shortTimestampWithTimeZoneReadFunction(),
+                            shortTimestampWithTimeZoneWriteFunction()));
+                }
         }
 
         if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
@@ -766,6 +785,23 @@ public class ClickHouseClient
             // ClickHouse stores incorrect results when the values are out of supported range.
             DATETIME.validate(version, timestamp);
             statement.setObject(index, timestamp);
+        };
+    }
+
+    private static LongReadFunction shortTimestampWithTimeZoneReadFunction()
+    {
+        return (resultSet, columnIndex) -> {
+            ZonedDateTime zonedDateTime = resultSet.getObject(columnIndex, ZonedDateTime.class);
+            return packDateTimeWithZone(zonedDateTime.toInstant().toEpochMilli(), zonedDateTime.getZone().getId());
+        };
+    }
+
+    private static LongWriteFunction shortTimestampWithTimeZoneWriteFunction()
+    {
+        return (statement, index, value) -> {
+            long millisUtc = unpackMillisUtc(value);
+            TimeZoneKey timeZoneKey = unpackZoneKey(value);
+            statement.setObject(index, Instant.ofEpochMilli(millisUtc).atZone(timeZoneKey.getZoneId()));
         };
     }
 

--- a/plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java
+++ b/plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java
@@ -33,6 +33,7 @@ import org.testng.annotations.Test;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.ZoneId;
+import java.util.function.Function;
 
 import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Verify.verify;
@@ -45,6 +46,7 @@ import static io.trino.spi.type.IntegerType.INTEGER;
 import static io.trino.spi.type.RealType.REAL;
 import static io.trino.spi.type.SmallintType.SMALLINT;
 import static io.trino.spi.type.TimestampType.createTimestampType;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_SECONDS;
 import static io.trino.spi.type.TinyintType.TINYINT;
 import static io.trino.spi.type.VarbinaryType.VARBINARY;
 import static io.trino.spi.type.VarcharType.VARCHAR;
@@ -778,6 +780,63 @@ public abstract class BaseClickHouseTypeMapping
         };
     }
 
+    @Test(dataProvider = "sessionZonesDataProvider")
+    public void testClickHouseDateTimeWithTimeZone(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        dateTimeWithTimeZoneTest(clickhouseDateTimeInputTypeFactory("datetime"))
+                .execute(getQueryRunner(), session, clickhouseCreateAndInsert("tpch.datetime_tz"));
+    }
+
+    private SqlDataTypeTest dateTimeWithTimeZoneTest(Function<ZoneId, String> inputTypeFactory)
+    {
+        ZoneId utc = ZoneId.of("UTC");
+        SqlDataTypeTest tests = SqlDataTypeTest.create()
+                .addRoundTrip(format("Nullable(%s)", inputTypeFactory.apply(utc)), "NULL", TIMESTAMP_TZ_SECONDS, "CAST(NULL AS TIMESTAMP(0) WITH TIME ZONE)")
+
+                // Since ClickHouse datetime(timezone) does not support values before epoch, we do not test this here.
+
+                // epoch
+                .addRoundTrip(inputTypeFactory.apply(utc), "0", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1970-01-01 00:00:00 Z'")
+                .addRoundTrip(inputTypeFactory.apply(utc), "'1970-01-01 00:00:00'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1970-01-01 00:00:00 Z'")
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'1970-01-01 00:00:00'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1970-01-01 05:30:00 +05:30'")
+
+                // after epoch
+                .addRoundTrip(inputTypeFactory.apply(utc), "'2019-03-18 10:01:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2019-03-18 10:01:17 Z'")
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'2019-03-18 10:01:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2019-03-18 10:01:17 +05:45'")
+                .addRoundTrip(inputTypeFactory.apply(ZoneId.of("GMT")), "'2019-03-18 10:01:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2019-03-18 10:01:17 Z'")
+                .addRoundTrip(inputTypeFactory.apply(ZoneId.of("UTC+00:00")), "'2019-03-18 10:01:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2019-03-18 10:01:17 Z'")
+
+                // time doubled in JVM zone
+                .addRoundTrip(inputTypeFactory.apply(utc), "'2018-10-28 01:33:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 01:33:17 Z'")
+                .addRoundTrip(inputTypeFactory.apply(jvmZone), "'2018-10-28 01:33:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 01:33:17 -05:00'")
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'2018-10-28 01:33:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 01:33:17 +05:45'")
+
+                // time doubled in Vilnius
+                .addRoundTrip(inputTypeFactory.apply(utc), "'2018-10-28 03:33:33'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 03:33:33 Z'")
+                .addRoundTrip(inputTypeFactory.apply(vilnius), "'2018-10-28 03:33:33'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 03:33:33 +03:00'")
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'2018-10-28 03:33:33'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-10-28 03:33:33 +05:45'")
+
+                // time gap in JVM zone
+                .addRoundTrip(inputTypeFactory.apply(utc), "'1970-01-01 00:13:42'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1970-01-01 00:13:42 Z'")
+                // TODO: Check the range of DateTime(timezone) values written from Trino to ClickHouse to prevent ClickHouse from storing incorrect results.
+                //       e.g. 1970-01-01 00:13:42 will become 1970-01-01 05:30:00
+                // .addRoundTrip(inputTypeFactory.apply(kathmandu), "'1970-01-01 00:13:42'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1970-01-01 00:13:42 +05:30'")
+                .addRoundTrip(inputTypeFactory.apply(utc), "'2018-04-01 02:13:55'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-04-01 02:13:55 Z'")
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'2018-04-01 02:13:55'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-04-01 02:13:55 +05:45'")
+
+                // time gap in Vilnius
+                .addRoundTrip(inputTypeFactory.apply(kathmandu), "'2018-03-25 03:17:17'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '2018-03-25 03:17:17 +05:45'")
+
+                // time gap in Kathmandu
+                .addRoundTrip(inputTypeFactory.apply(vilnius), "'1986-01-01 00:13:07'", TIMESTAMP_TZ_SECONDS, "TIMESTAMP '1986-01-01 00:13:07 +03:00'");
+
+        return tests;
+    }
+
     @DataProvider
     public Object[][] sessionZonesDataProvider()
     {
@@ -883,5 +942,10 @@ public abstract class BaseClickHouseTypeMapping
     protected SqlExecutor onRemoteDatabase()
     {
         return clickhouseServer::execute;
+    }
+
+    private static Function<ZoneId, String> clickhouseDateTimeInputTypeFactory(String inputTypePrefix)
+    {
+        return zone -> format("%s('%s')", inputTypePrefix, zone);
     }
 }

